lib/queuePopulator/KafkaLogConsumer/LogConsumer.js

```diff
@@ -153,18 +248,80 @@ class LogConsumer {
      */
     readRecords(params, cb) {
         async.series([
-            // waiting for the consumer group to rebalance
-            next => this._waitForAssignment(0, next),
-            // saving last commited offset
-            next => this._storeCurrentOffsets(next),
             // consuming the desired number of messages at most
             next => this._consumeKafkaMessages(params.limit, next),
             next => {
                 // ending and returning the stream
                 this._listRecordStream.end();
                 return next(null, { log: this._listRecordStream, tailable: false });
             }
-        ], (err, res) => cb(err, res[3]));
+        ], (err, res) => cb(err, res?.[1]));
+    }
+
+    /**
+     * Stores the offsets locally for the consumer to
+     * auto commit them at a later time
+     * @returns {undefined}
+     */
+    storeOffsets() {
+        if (!this._topicPartition || this._topicPartition.length === 0) {
+            this._log.debug('No offsets to store', {
+                method: 'LogConsumer.storeOffsets',
+                topic: this._topic,
+                consumerGroupId: this._consumerGroupId,
+            });
+            return;
+        }
+        this._consumer.offsetsStore(this._topicPartition);
+        this._log.info('Offsets stored', {
+            method: 'LogConsumer.storeOffsets',
+            consumerGroupId: this._consumerGroupId,
+            topicPartition: this._topicPartition,
+        });
+        this._pendingCommit = true;
+        this._topicPartition = null;
+    }
+
+    /**
+     * LogConsumer is considered ready if it is connected
+     * to the Kafka broker and ready to consume messages.
+     * @returns {boolean}
+     */
+    isReady() {
+        return this._consumer.isConnected();
+    }
+
+    /**
+     * Closes the consumer connection, while making sure
+     * that all messages are processed and offsets are committed.
+     * @param {Function} cb
+     */
+    close(cb) {
+        async.series([
+            next => {
+                if (this._consumer?.isConnected()) {
+                    const subscription = this._consumer.subscription() || [];
+                    if (subscription.length > 0) {
+                        // we first unsubscribe from the topic
+                        // to initiate the rebalance process
+                        // allowing the message processing to finish
+                        // and offsets to be committed
+                        this._consumer.unsubscribe();
+                        this.once('unassigned', () => next());
+                        return null;
+                    }
+                }
+                return next();
+            },
+            next => {
+                if (this._consumer?.isConnected()) {
+                    this._consumer.disconnect();
+                    this._consumer.once('disconnected', () => next());
+                    return null;
```

Comment:
should we not do it the other way, to avoid race conditions?
(node is thread safe, so probably fine as is... but could change with futurechanges...)
```suggestion                  
                    this._consumer.once('disconnected', () => next());
                    this._consumer.disconnect();
                    return null;
```

(unless the semantics is that setting `once('disconnected')` callback will immediately invoke the callback if the consumer is alreadyd disconnected)